### DISCUSSION TITLE
Use Common Files in all Components

### DISCUSSION THREAD

**Discussion by sheershak-agarwal** [nasa/fprime#940] [Q&A]
Hello,

I want to use some common c++ files in all my components.

What I did:
1. created a new folder common
2. add my cpp and h files in the common folder
3. added CmakeLists.txt in the Common folder saying 
set(SOURCE_FILES
	"${CMAKE_CURRENT_LIST_DIR}/Common.cpp"
)
register_fprime_module()
4. add #include "App/Common/Common.h" on top of ComponentImpl.cpp
5. added add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/Common/") in the Main CmakeLists.txt

but when I build it cannot find the Common.h file.
I have also tried it without the CmakeList.txt file. How can add a file in a common folder and then add use those in my ComponentImpl design?

**Comment by timcanham** [comment]
The header file dependencies don't work quite right yet, so you have to add:

```cmake
set(MOD_DEPS
  Path/ToLibrary
)
```

To your `CmakeList.txt` for the component(s) that depend on it.

**Comment by sheershak-agarwal** [comment]
I added to my component
set(MOD_DEPS
  "App/Common/StateMachine.cpp"
)
but I see this error:
CMake Error at /home/sheershak/Desktop/F-Prime-Debug/fprime/cmake/support/Module.cmake:94 (add_dependencies):
-- Generating done
  **The dependency target "App_Common_StateMachine.cpp" of target**
  "App_MissionStateManager" does not exist.
Call Stack (most recent call first):
  /home/sheershak/Desktop/F-Prime-Debug/fprime/cmake/support/Module.cmake:146 (generate_module)
  /home/sheershak/Desktop/F-Prime-Debug/fprime/cmake/API.cmake:182 (generate_library)
  MissionStateManager/CMakeLists.txt:9 (register_fprime_module)

**Reply by Joshua-Anderson** [reply]
MOD_DEPS will be the folder name of the module you're depending on, give `"App/Common"` a try.

**Reply by sheershak-agarwal** [reply]
Still getting this error:

CMake Error at /home/sheershak/Desktop/F-Prime-Debug/fprime/cmake/support/Module.cmake:94 (add_dependencies):
-- Generating done
  The dependency target "App_Common" of target "App_MissionStateManager" does
  not exist.
Call Stack (most recent call first):
  /home/sheershak/Desktop/F-Prime-Debug/fprime/cmake/support/Module.cmake:146 (generate_module)
  /home/sheershak/Desktop/F-Prime-Debug/fprime/cmake/API.cmake:182 (generate_library)
  MissionStateManager/CMakeLists.txt:10 (register_fprime_module)

**Reply by Joshua-Anderson** [reply]
Do you have a CMakeLists.txt file defining a component in App/Common?

**Reply by sheershak-agarwal** [reply]
No(its not a component but just c++ files), so should I add one and should I also added add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/Common/") in the Main CmakeLists.txt

**Reply by Joshua-Anderson** [reply]
Yep, exactly. Add a CMakeLists.txt file with the cpp file as a source file and declaring a module, then add it to the parent directory's CMakeLists.txt file.

**Reply by sheershak-agarwal** [reply]
Thank you for that. It compiles perfectly now.

**Comment by timcanham** [comment]
Sorry - to clarify, you put the directory in the `MOD_DEPS`. For you:

```cmake
set(MOD_DEPS
  App/Common
)
```

**Reply by sheershak-agarwal** [reply]
I still get an error:
CMake Error at /home/sheershak/Desktop/F-Prime-Debug/fprime/cmake/support/Module.cmake:94 (add_dependencies):
-- Generating done
  The dependency target "App_Common" of target "App_MissionStateManager" does
  not exist.
Call Stack (most recent call first):
  /home/sheershak/Desktop/F-Prime-Debug/fprime/cmake/support/Module.cmake:146 (generate_module)
  /home/sheershak/Desktop/F-Prime-Debug/fprime/cmake/API.cmake:182 (generate_library)
  MissionStateManager/CMakeLists.txt:10 (register_fprime_module)

**Comment by timcanham** [comment]
There is probably something off in the pathnames. Can you show your directory structure and where the cmake files are?

**Reply by sheershak-agarwal** [reply]
https://github.com/sheershak-agarwal/F-Prime-Debug
and the edit I made is in https://github.com/sheershak-agarwal/F-Prime-Debug/blob/main/App/MissionStateManager/CMakeLists.txt

**Comment by LeStarch** [comment]
I may be late to the game here but it looks like the steps you are following are ok.

1. Add CMakeLists to App/Common
2. Add MOD_DEP of "App/Common" to MissionStateManager
3. (Did you do this) Add "add_fprime_subdirectory(/path/to/common)" to "App/CMakeLists.txt"
